### DISCUSSION TITLE
Font not loading inside PDF

### DISCUSSION THREAD

**Discussion by Rpgdudester** [dompdf/dompdf#2857] [Q&A]
Good evening,

First I want to say I love Dompdf and I hope sooner or later it gets bootstrap 4/5 and css3 support but that's another matter.

The problem I'm having right now is i'm trying to use a font called Autography and because it's not a google font, it's being a pain in the rear trying to load it. 

Things I've done so far:

I've loaded the font into the dompdf font directory using load_font.pdf and checked the dompdf_font_family_cache.php file to make sure it loaded the files in there.

<img width="515" alt="Screen Shot 2022-04-15 at 5 06 35 PM" src="https://user-images.githubusercontent.com/44619542/163646015-e185dfff-7d90-4175-ba20-de9dd356a9d2.png">

<img width="528" alt="image" src="https://user-images.githubusercontent.com/44619542/163646317-43e83c4c-a7a6-4d99-be3d-7d187065ce96.png">

and made sure that the file was a ttf file.

<img width="175" alt="image" src="https://user-images.githubusercontent.com/44619542/163648455-8b9b6764-7a73-48b9-8e32-5679a49252a5.png">

and also here's my directory setup if you need to figure out why i did stuff the way I did :p 

the ending result is the font falls back to the backup font and doesn't seem to want to load. My bosses paid for the font use for commercial use so I'd like to be able to get it working.

Everything else loads up great, using bootstrap 3.3.7 i was able to get a shipping manifest built and everything, it's just the last part of this project is getting 1 paragraph changed to that font. and i've been racking my brain pouring over google trying to find the answer. and nothing I have found seems to help so far so I'm taking it to the pros!

so if you have any suggestions or ideas, please let me know.

- Joshua

**Comment by bsweeney** [comment]
If you've correctly loaded the fonts then you shouldn't need to use `@font-face`. It's been a while since I've done any work on that script so it maybe doesn't work as well as it used to.

Are you loading the HTML into Dompdf using `loadHtml`? This is probably where the `@font-face` declaration is failing. Your path is absolute and file-based paths are absolute from the _root of the filesystem_. Try a relative path and see if that helps. Or an absolute path from the filesystem root. I'm looking to add a "document root" option (possibly in 2.0.0) to help alleviate this type of issue.

Lastly, how are you applying the font to your document? You showed the `@font-face` part but not how you're using it in your document.

**Comment by Rpgdudester** [comment]
1st answer: I'm writing everything in an extended php variable then using loadHtml to load the single variable which works out great except for that single font issue.

2nd answer:
<img width="412" alt="Screen Shot 2022-04-15 at 8 49 16 PM" src="https://user-images.githubusercontent.com/44619542/163657109-986926e2-280d-4401-9587-0e4d84fb857b.png">

then assigning the class. 

I've tried with both absolute and relative locations and so far both have not worked. they both seem to fall back to the backup font.

If you're telling me that i'm doing everything right from your view, then it's probably something stupid on my side.
 

and I just tried assigning a font-weight both of normal and 400 and both of them make the pdf not create (I create the pdf file then i echo the variable to see if something is missing) so that may point to something.

**Reply by bsweeney** [reply]
As for the relative location, since I didn't mention it, that would be relative the the PHP script utilizing Dompdf.

**Comment by Rpgdudester** [comment]
Good morning Brian,

I was up till about 1 AM last night playing with my code and fighting with the font. I commented out most of my code except for 3 lines, then tried to add that font (as a class) to the html element, nothing. 

I even went so far as to rebuilt the cache for the fonts. still nothing has worked. 

I also just tried to change the permissions on dompdf to 775 just to make sure it wasn't a permissions issue and it still didn't load the font. 

When I set the temp directory to my temp folder, I can see it creating the logo png temp file but not the font one.

I have 2 other fonts loading from google fonts which work perfectly, just not this one.

i'm not sure what else to try. 

You're saying it should be working fine, either with @font-face or actually adding it into loading it into dompdf with the font_load.php file and it's not, so either i'm loading it wrong, or it's being loaded into dompdf wrong. (it's probably me)


The one thing I did notice (Though I could be wrong) is that the when I check the dompdf_font_familt_cache.php against the dompdf_font_familt_cache.dist.php, one is using brackets as its array, and one is using braces, do you think maybe when using font_load.php, it's loading my font info into the wrong file?

Just wondering.

Thanks!

**Reply by bsweeney** [reply]
Since the font file is local it shouldn't create a temp file. The font, when loaded, should go directly into your font directory. I didn't notice that you weren't setting the fontDir path. If that were the issue changing the permissions would have addressed it. Still, you might want to set that option so you're not writing files to the Dompdf dir.

The array format shouldn't impact font loading, so long as the cache file contains valid PHP.

**Comment by Rpgdudester** [comment]
Ok, So i've been playing with this off and on all day. and I've come up with 2 reasons.

1. it's because the font file is coming in as a OTF file so I have to convert it externally to a TTF file then install the font with it in which case, where ever i'm converting it from otf to ttf is doing it wrong.
2. the ttf file itself is having problems loading in DOMpdf, neither of which are DOMpdf's issue, but mine.

As I said, Dom's a great pdf maker so i'm going to keep digging at this and maybe i'll find something. but I'm considering the issue as being on my side more than anything else now.

Thanks for your help!

**Reply by bsweeney** [reply]
I didn't bring up OTF as a potential issue since your logic referenced a TTF file. But yes, our font parser (php-font-lib) does not yet support OTF fonts.

**Reply by Rpgdudester** [reply]
Right, I understand that, 

I used an online converter that converts an OTF to a TTF. Used the load_font.php file to load the TTF, it creates the ufm and umf.php file for the font in the fonts directory, i set the font-family and font-weight on an h6 element, (using only the h6 as a test). 

After doing this, it fails to create the pdf document.

If i remove the font-weight, it creates the document with the sans-serif as the backup font. so it leads me to believe that it's got a problem with the font or font-weight itself. the umf.php says isUnicode => true, so i wonder if maybe it's a unicode issue then?

**Reply by Rpgdudester** [reply]
This is going to sound stupid sir, but is there a debug or something i can set an option for so I can run it and see if it's a permissions issue, a folder issue, etc?

**Comment by Rpgdudester** [comment]
Ok, I just did a debug in my IDE and literally followed through the code and saw this:

<img width="1200" alt="image" src="https://user-images.githubusercontent.com/44619542/163732267-8fbd41c1-da57-4541-a43a-1fc479f3afb3.png">


So this explains ALOT on why it wouldn't load the font in the editor!

So we can consider this issue fixed then. it's definitely a my side issue more than anything. Thank you for all your help Brian!!

**Comment by Rpgdudester** [comment]
After getting the chroot and the directories set right for the fonts and such, i was able to get it passed the permission denied hump but now i've run into something even more amusing. 

<img width="339" alt="image" src="https://user-images.githubusercontent.com/44619542/163734925-327dfbae-eea0-4074-aa7d-a12d641a24ae.png">

I checked the FontLib\Table\Type folder and I don't have a fftm.php file. so it's trying to load from my auto loader for my site and of course it won't find it in my auto loader, so now it's failing crashing because it can't find that table. 

I checked the the php-font-lib section here: https://github.com/dompdf/php-font-lib/tree/master/src/FontLib/Table/Type just to make sure I wasn't missing any files and I didn't see a table type there for that name either. 

https://www.fontforestry.com/autography-font-free-download/ This is where I got the file from if you want to do any testing with that font.

Did I mention I love DOMpdf and I appreciate all the hard work you put into this code? :)

-Joshua

**Reply by bsweeney** [reply]
Thanks for your appreciation. It's ... appreciated! 😄 

As for the error regarding the FFTM table, we have run into it before (see dompdf/php-font-lib#48). I believe it's been addressed. What version of php-font-lib are you running?

**Reply by Rpgdudester** [reply]
ok, I just checked and it's using whatever version you have in the domPDF composer.json, so 0.5.4 is the one i'm using.

I just did a check and the same issue that was fixed here: https://github.com/dompdf/php-font-lib/commit/04b2cf6d437eedc1089f5c233c81f1b5b2e67622 is actually the same exact thing i have in my file

<img width="719" alt="image" src="https://user-images.githubusercontent.com/44619542/163743155-8c33d894-9c9f-4fde-978a-5e1075f0a3b2.png">

so I may have to take the advice the link you posted in the link and create a fftm.php file and see what i can come up with then.

**Comment by Rpgdudester** [comment]
I am so happy to announce, after fighting with this for 2-3 days, I finally got the font to show up right on the pdf file, @bsweeney, you pointed me in the right direction with having to build that ttfm table, then a couple other tables showed up so i did the same thing. I also took the AFM file that i converted with the ttf file, and copy/pasted that as a umf file (it looks like the AFM vs UMF stuff isn't copying over right) and tried it again and everything magically worked.

Thank you so much Brian!
